scan.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"fmt"
+	"hash/fnv"
+	"io"
+	"log"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	chunker "github.com/ipfs/go-ipfs-chunker"
+)
+
+type fileChunk struct {
+	fileid int
+	offset int64
+	length int64
+	sum    uint64
+}
+
+func ScanFiles(filelist []string) error {
+	var wait sync.WaitGroup
+
+	lookup := make(map[uint64][]fileChunk)
+	var lookupLk sync.Mutex
+
+	jobs := make(chan int)
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		go func() {
+			for fid := range jobs {
+				file := filelist[fid]
+				chunks, err := scanFile(fid, file)
+				if err != nil {
+					log.Printf("failed to hash file %q: %s", file, err)
+				}
+
+				lookupLk.Lock()
+				for _, chunk := range chunks {
+					lookup[chunk.sum] = append(lookup[chunk.sum], chunk)
+				}
+				lookupLk.Unlock()
+				wait.Done()
+			}
+
+		}()
+	}
+
+	hashStart := time.Now()
+	for f := range filelist {
+		wait.Add(1)
+		jobs <- f
+	}
+	close(jobs)
+
+	wait.Wait()
+	log.Printf("Hashing data took %s", time.Since(hashStart))
+
+	for _, dupes := range lookup {
+		if len(dupes) <= 0 {
```

Comment:
I would put this as 1 isntead of zero, no point printing out chunks without a pair

Reply:
Good catch!